src/main/java/net/minestom/server/inventory/InventoryImpl.java

```diff
@@ -0,0 +1,309 @@
+        final ItemStack currentCursorItem = cursorPlayersItem.getOrDefault(player, ItemStack.AIR);
+        if (!currentCursorItem.equals(cursorItem)) {
+            player.sendPacket(SetSlotPacket.createCursorPacket(cursorItem));
+        }
+    }
+
+    @Override
+    public void update() {
+        this.viewers.forEach(this::update);
+    }
+
+    @Override
+    public void update(@NotNull Player player) {
+        player.sendPacket(new WindowItemsPacket(getWindowId(), 0, List.of(itemStacks), cursorPlayersItem.getOrDefault(player, ItemStack.AIR)));
+    }
+
+    @Override
+    public @Nullable ClickResult handleClick(@NotNull Player player, @NotNull ClickInfo clickInfo) {
+        return DEFAULT_HANDLER.handleClick(this, player, clickInfo);
+    }
+
+    @Override
+    public @NotNull ItemStack getItemStack(int slot) {
+        return (ItemStack) ITEM_UPDATER.getVolatile(itemStacks, slot);
+    }
+
+    @Override
+    public @NotNull ItemStack[] getItemStacks() {
+        return itemStacks.clone();
+    }
+
+    @Override
+    public void copyContents(@NotNull ItemStack[] itemStacks) {
+        Check.argCondition(itemStacks.length != getSize(),
+                "The size of the array has to be of the same size as the inventory: " + getSize());
+
+        for (int i = 0; i < itemStacks.length; i++) {
+            final ItemStack itemStack = itemStacks[i];
+            Check.notNull(itemStack, "The item array cannot contain any null element!");
+            setItemStack(i, itemStack);
+        }
+    }
+
+    @Override
+    public synchronized void setItemStack(int slot, @NotNull ItemStack itemStack) {
+        Check.argCondition(!MathUtils.isBetween(slot, 0, getSize()),
+                "Inventory does not have the slot " + slot);
+        safeItemInsert(slot, itemStack);
+    }
+
+    protected final void safeItemInsert(int slot, @NotNull ItemStack itemStack) {
+        safeItemInsert(slot, itemStack, true);
+    }
+
+    /**
+     * Inserts safely an item into the inventory.
+     * <p>
+     * This will update the slot for all viewers and warn the inventory that
+     * the window items packet is not up-to-date.
+     *
+     * @param slot      the internal slot id
+     * @param itemStack the item to insert (use air instead of null)
+     * @throws IllegalArgumentException if the slot {@code slot} does not exist
+     */
+    protected final void safeItemInsert(int slot, @NotNull ItemStack itemStack, boolean sendPacket) {
+        ItemStack previous;
+        synchronized (this) {
+            previous = itemStacks[slot];
+            if (itemStack.equals(previous)) return; // Avoid sending updates if the item has not changed
+
+            UNSAFE_itemInsert(slot, itemStack);
+            if (sendPacket) updateSlot(slot, itemStack);
+        }
+
+        EventDispatcher.call(new InventoryItemChangeEvent(this, slot, previous, itemStack));
+    }
+
+    protected void UNSAFE_itemInsert(int slot, @NotNull ItemStack itemStack) {
```

Comment:
What is the justification for this existing?

Reply:
this is behaviour carried over from Minestom's previous inventory system. Currently it's just used for allowing PlayerInventory to call EntityEquipEvent if necessary. I wasn't really sure how to remove it without removing EntityEquipEvent as a whole, which I have no real reason to do